src/main/java/org/apache/commons/compress/archivers/tar/TarFile.java

```diff
@@ -0,0 +1,736 @@
+            while ((length = in.read(smallBuf)) >= 0) {
+                longName.write(smallBuf, 0, length);
+            }
+        }
+        getNextTarEntry();
+        if (currEntry == null) {
+            // Bugzilla: 40334
+            // Malformed tar file - long entry name not followed by entry
+            return null;
+        }
+        byte[] longNameData = longName.toByteArray();
+        // remove trailing null terminator(s)
+        length = longNameData.length;
+        while (length > 0 && longNameData[length - 1] == 0) {
+            --length;
+        }
+        if (length != longNameData.length) {
+            final byte[] l = new byte[length];
+            System.arraycopy(longNameData, 0, l, 0, length);
+            longNameData = l;
+        }
+        return longNameData;
+    }
+
+    /**
+     * The last record block should be written at the full size, so skip any
+     * additional space used to fill a record after an entry
+     *
+     * @throws IOException when skipping the padding of the record fails
+     */
+    private void skipRecordPadding() throws IOException {
+        if (!isDirectory() && currEntry.getSize() > 0 && currEntry.getSize() % recordSize != 0) {
+            final long numRecords = (currEntry.getSize() / recordSize) + 1;
+            final long padding = (numRecords * recordSize) - currEntry.getSize();
+            archive.position(archive.position() + padding);
+            throwExceptionIfPositionIsNotInArchive();
+        }
+    }
+
+    /**
+     * Checks if the current position of the SeekableByteChannel is in the archive.
+     * @throws IOException If the position is not in the archive
+     */
+    private void throwExceptionIfPositionIsNotInArchive() throws IOException {
+        if (archive.size() < archive.position()) {
+            throw new IOException("Truncated TAR archive");
+        }
+    }
+
+    /**
+     * Get the next record in this tar archive. This will skip
+     * over any remaining data in the current entry, if there
+     * is one, and place the input stream at the header of the
+     * next entry.
+     *
+     * <p>If there are no more entries in the archive, null will be
+     * returned to indicate that the end of the archive has been
+     * reached.  At the same time the {@code hasHitEOF} marker will be
+     * set to true.</p>
+     *
+     * @return The next TarEntry in the archive, or null if there is no next entry.
+     * @throws IOException when reading the next TarEntry fails
+     */
+    private ByteBuffer getRecord() throws IOException {
+        ByteBuffer headerBuf = readRecord();
+        setAtEOF(isEOFRecord(headerBuf));
+        if (isAtEOF() && headerBuf != null) {
+            // Consume rest
+            tryToConsumeSecondEOFRecord();
+            // TODO: This is present in the TarArchiveInputStream but I don't know if we need this in the random access implementation. All tests are passing ...
+            // consumeRemainderOfLastBlock();
```

Comment:
I think we should leave the `consumeRemainderOfLastBlock` as it was.

According to the comment :
```
    /**
     * This method is invoked once the end of the archive is hit, it
     * tries to consume the remaining bytes under the assumption that
     * the tool creating this archive has padded the last block.
     */
```

This is used to deal with the last block with padding zero. Even through some tar implementions don't implement this, this is part of tar specifiction.

See also [this](https://www.gnu.org/software/tar/manual/html_node/Standard.html) : 

> When writing an archive, the last record of blocks should be written at the full size, with blocks after the zero block containing all zeros.

Reply:
I added the implementation. The test which checks this for the `TarArchiveInputStream` was a bit obscured. I added a bit of documentation so the next one finds it easier.